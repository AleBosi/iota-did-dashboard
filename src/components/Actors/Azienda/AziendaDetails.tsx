import React, { useMemo, useState } from "react";
import UnlockAccountModal from "@/components/Common/UnlockAccountModal";
import SeedPasswordModal from "@/components/Common/SeedPasswordModal";
import { useSecrets } from "@/contexts/SecretsContext";
import { hasEncryptedSeed, generateMnemonic24, deriveMockAccount } from "@/utils/cryptoUtils";

export default function AziendaDetails({
  azienda,
  credits = 0,
  onUpdate,
  onDelete,
  showSecrets, // ignorato: la seed va mostrata solo dopo sblocco
}: {
  azienda: any;
  credits?: number;
  onUpdate: (a: any) => void;
  onDelete: () => void;
  showSecrets?: boolean;
}) {
  const a = azienda || {};
  const li = a.legalInfo || {};
  const did = (a.did || a.id || "").trim();

  const { getSeed, setSeed, clearSeed } = useSecrets();
  const unlocked = getSeed({ type: "company", id: did });

  const seedExists = useMemo(() => (did ? hasEncryptedSeed("company", did) : false), [did]);

  const [showUnlock, setShowUnlock] = useState(false);

  // per aziende legacy senza seed salvata: generiamo una mnemonic e apriamo la modale di salvataggio
  const [setPwdOpen, setSetPwdOpen] = useState(false);
  const [mnemonicToSave, setMnemonicToSave] = useState<string>("");

  const copy = (txt: string) => {
    try {
      navigator.clipboard.writeText(txt);
      alert("Copiato negli appunti");
    } catch {}
  };

  const handleSetPassword = () => {
    // 1) se l’azienda ha già una seed valida (>=12 parole), riusiamola; altrimenti generiamo
    const existing =
      typeof a.seed === "string" && a.seed.split(" ").length >= 12 ? a.seed : "";

    const m = existing || generateMnemonic24(); // 24 parole reali BIP39
    setMnemonicToSave(m);

    // 2) allinea DID/Address alla seed (stessa pipeline di AziendaForm)
    try {
      const acc = deriveMockAccount(m);
      const newDid = `did:iota:evm:${acc.address}`;
      if (did !== newDid) {
        onUpdate({ ...a, did: newDid, evmAddress: acc.address, seed: a.seed || m });
      }
    } catch {
      // non bloccare l'UI in caso di errore di derivazione
    }

    // 3) apri la modale di impostazione password/salvataggio cifrato
    setSetPwdOpen(true);
  };

  // estrai address da DID iota:evm se presente (utile da mostrare)
  const address =
    did.startsWith("did:iota:evm:") ? (did.split(":").pop() as string) : undefined;

  return (
    <div>
      <div className="space-y-1 text-sm">
        <div>
          <strong>Ragione sociale:</strong> {a.name}
        </div>

        <div>
          <strong>DID:</strong>{" "}
          <code className="text-xs">{did || "-"}</code>{" "}
          {did && (
            <button className="text-blue-600 underline" onClick={() => copy(did)}>
              Copia
            </button>
          )}
        </div>

        {address && (
          <div>
            <strong>Address:</strong>{" "}
            <code className="text-xs">{address}</code>{" "}
            <button className="text-blue-600 underline" onClick={() => copy(address)}>
              Copia
            </button>
          </div>
        )}

        {/* ✅ Seed mostrata solo se sbloccata */}
        {unlocked && (
          <div className="mt-2">
            <strong>Seed (24 parole):</strong>
            <div className="mt-1 rounded border bg-gray-50 p-2 text-xs leading-6 select-all">
              {unlocked}
            </div>
          </div>
        )}

        <div>
          <strong>P.IVA:</strong> {li.vat || "-"}
        </div>
        <div>
          <strong>LEI:</strong> {li.lei || "-"}
        </div>
        <div>
          <strong>Indirizzo:</strong> {li.address || "-"}
        </div>
        <div>
          <strong>Email:</strong> {li.email || "-"}
        </div>
        <div>
          <strong>Nazione:</strong> {li.country || "-"}
        </div>
        <div>
          <strong>Crediti azienda:</strong> {credits.toLocaleString()}
        </div>
        <div>
          <strong>Creato il:</strong> {a.createdAt}
        </div>
      </div>

      <div className="mt-4 flex flex-wrap gap-2">
        <button
          className="rounded px-3 py-1 border hover:bg-gray-100"
          onClick={() => onUpdate({ ...a })}
        >
          Salva modifiche
        </button>
        <button
          className="rounded px-3 py-1 border border-red-300 text-red-700 hover:bg-red-50"
          onClick={onDelete}
        >
          Elimina
        </button>

        {/* Azioni seed */}
        {!seedExists && did && (
          <button
            className="rounded px-3 py-1 border bg-amber-600 text-white hover:opacity-90"
            onClick={handleSetPassword}
          >
            Imposta password & salva seed
          </button>
        )}

        {seedExists && !unlocked && (
          <button
            className="rounded px-3 py-1 border bg-gray-900 text-white hover:opacity-90"
            onClick={() => setShowUnlock(true)}
          >
            Sblocca e mostra seed
          </button>
        )}

        {unlocked && (
          <button
            className="rounded px-3 py-1 border hover:bg-gray-100"
            onClick={() => clearSeed({ type: "company", id: did })}
          >
            Nascondi seed
          </button>
        )}
      </div>

      {/* Modal sblocco (password -> seed in chiaro in RAM) */}
      <UnlockAccountModal
        open={showUnlock}
        onClose={() => setShowUnlock(false)}
        entityType="company"
        entityId={did}
        entityName={a.name}
        onUnlocked={(mnemonic) => setSeed({ type: "company", id: did }, mnemonic)}
      />

      {/* Modal set password & salva (per aziende legacy senza seed salvata) */}
      <SeedPasswordModal
        open={setPwdOpen}
        onClose={() => setSetPwdOpen(false)}
        entityType="company"
        entityId={did}
        entityName={a.name}
        mnemonic={mnemonicToSave}
        onSaved={() => {
          // una volta salvata, possiamo anche sbloccarla subito in RAM per comodità
          setSeed({ type: "company", id: did }, mnemonicToSave);
          setSetPwdOpen(false);
        }}
      />
    </div>
  );
}
