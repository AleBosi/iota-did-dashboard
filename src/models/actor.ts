// /models/actor.ts

export interface Actor {
  id: string; // DID unico (es: "did:iota:evm:0x...")
  name: string;
  role: "admin" | "azienda" | "creator" | "operatore" | "macchinario";
  aziendaId?: string;      // DID dell’azienda di appartenenza
  seed?: string;           // non esportare mai in VC!
  publicKey?: string;      // per interoperabilità/validazione
  createdAt?: string;      // opzionale, ISO date
  updatedAt?: string;      // opzionale

  // AGGIUNGI QUESTO CAMPO:
  vcIds?: string[];        // Lista degli ID delle VC associate a questo attore
}
