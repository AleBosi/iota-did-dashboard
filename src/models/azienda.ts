// /models/azienda.ts
export interface Azienda {
  id: string;         // DID (es: did:iota:evm:...)
  name: string;
  seed: string;       // per login e firma
  legalInfo?: {
    vat?: string;         // partita IVA o VAT
    lei?: string;         // Legal Entity Identifier (EBSI ready)
    address?: string;
    email?: string;
    country?: string;
    [key: string]: any;   // altri campi liberi
  };
  creators: string[];     // array di DID
  operatori: string[];
  macchinari: string[];
  createdAt?: string;     // ISO
  updatedAt?: string;     // ISO
}
