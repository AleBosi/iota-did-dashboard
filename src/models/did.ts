// /models/did.ts
export interface DidDocument {
  id: string; // es: "did:iota:evm:0x123..."
  controller?: string[];
  verificationMethod?: {
    id: string;
    type: string;
    publicKeyMultibase: string;
  }[];
  authentication?: string[];
  service?: {
    id: string;
    type: string;
    serviceEndpoint: string;
  }[];
  // Altri campi secondo specifica, se serve
}
