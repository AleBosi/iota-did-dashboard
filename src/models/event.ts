// /models/event.ts

export interface Event {
  id: string;
  productId: string;         // ID prodotto collegato (ora obbligatorio)
  operatoreId: string;       // ID operatore coinvolto
  macchinarioId: string;     // ID macchinario coinvolto
  type: string;              // es: "Assemblaggio", "Controllo", "Assegnazione"
  description: string;
  date: string;              // ISO
  creatorId: string;         // DID del creator che ha creato lâ€™evento
  done?: boolean;            // true se eseguito
  bomComponent?: string;     // opzionale, componente specifico
  proofId?: string;          // se firmerai evento come VC
  vcIds?: string[];          // VC associate/notarizzate collegate a questo evento
}
