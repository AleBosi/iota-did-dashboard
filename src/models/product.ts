// /models/product.ts
export interface Product {
  id: string;              // ID univoco locale/app
  did: string;             // DID prodotto (es: did:iota:evm:0x...)
  name: string;            // Nome commerciale o etichetta prodotto
  typeId?: string;         // riferimento a ProductType.id
  serial?: string;         // seriale prodotto
  owner?: string;          // DID del proprietario corrente
  bom?: { id: string; name: string }[]; // lista componenti (Bill of Materials)
  credentials?: string[];  // lista di VC associate (ID delle VC)
  description?: string;    // opzionale
}
