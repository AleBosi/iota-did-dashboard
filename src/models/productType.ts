export interface ProductType {
  id: string;               // es: GTIN, oppure UUID univoco
  name: string;
  description?: string;
  // Se vuoi tipizzare la proprietà extra, puoi aggiungerla così:
  // standardCertVCId?: string;
}
