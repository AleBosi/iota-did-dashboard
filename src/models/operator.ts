// /models/operator.ts
import { Actor } from "./actor";

export interface Operator extends Actor {
  // aggiungi campi solo se servono SOLO agli operatori
}
