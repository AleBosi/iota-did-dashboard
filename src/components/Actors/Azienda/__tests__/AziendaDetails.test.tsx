import { render, screen } from "@testing-library/react";
import { Azienda } from "../../../../models/azienda"
import AziendaDetails from "../AziendaDetails";

const sampleAzienda: Azienda = {
  id: "did:iota:evm:1234",
  name: "Azienda Test",
  seed: "SEED1234",
  legalInfo: {
    vat: "12345678901",
    lei: "LEI123456",
    address: "Via Roma, 1",
    email: "info@azienda.it",
    country: "Italia",
  },
  creators: ["did:iota:creator1"],
  operatori: ["did:iota:op1"],
  macchinari: ["did:iota:mac1"],
  createdAt: "2025-01-01T12:00:00Z",
};

describe("AziendaDetails", () => {
  test("renderizza correttamente tutti i dati", () => {
    render(<AziendaDetails azienda={sampleAzienda} />);
    
    // Etichette (nessun errore se ci sono duplicati)
    expect(screen.getAllByText(/ragione sociale:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/did:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/seed:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/p\.iva:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/lei:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/indirizzo:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/email:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/nazione:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/creatori:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/operatori:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/macchinari:/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/creato il:/i).length).toBeGreaterThan(0);

    // Valori univoci
    expect(screen.getByText(sampleAzienda.name)).toBeInTheDocument();
    expect(screen.getByText(sampleAzienda.id)).toBeInTheDocument();
    expect(screen.getByText(sampleAzienda.seed)).toBeInTheDocument();
    expect(screen.getByText(sampleAzienda.legalInfo!.vat!)).toBeInTheDocument();
    expect(screen.getByText(sampleAzienda.legalInfo!.lei!)).toBeInTheDocument();
    expect(screen.getByText(sampleAzienda.legalInfo!.address!)).toBeInTheDocument();
    expect(screen.getByText(sampleAzienda.legalInfo!.email!)).toBeInTheDocument();
    expect(screen.getByText(sampleAzienda.legalInfo!.country!)).toBeInTheDocument();

    // Numeri accanto alle etichette
    expect(screen.getByText(/creatori:/i).parentElement).toHaveTextContent("1");
    expect(screen.getByText(/operatori:/i).parentElement).toHaveTextContent("1");
    expect(screen.getByText(/macchinari:/i).parentElement).toHaveTextContent("1");

    // Data robusta (matcher che ritorna sempre boolean)
    if (sampleAzienda.createdAt) {
      const nodes = screen.getAllByText((_, node) =>
        Boolean(node && node.textContent && node.textContent.includes(sampleAzienda.createdAt!))
      );
      expect(nodes.length).toBeGreaterThan(0);
    }
  });
  test("matcha lo snapshot", () => {
    const { asFragment } = render(<AziendaDetails azienda={sampleAzienda} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
