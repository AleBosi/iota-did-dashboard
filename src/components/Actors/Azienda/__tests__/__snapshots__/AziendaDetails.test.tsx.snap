// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AziendaDetails matcha lo snapshot 1`] = `
<DocumentFragment>
  <div
    class="border rounded p-4 bg-gray-50 mb-2"
  >
    <div>
      <b>
        Ragione sociale:
      </b>
       Azienda Test
    </div>
    <div>
      <b>
        DID:
      </b>
       did:iota:evm:1234
    </div>
    <div>
      <b>
        Seed:
      </b>
       
      <code>
        SEED1234
      </code>
    </div>
    <div>
      <b>
        P.IVA:
      </b>
       12345678901
    </div>
    <div>
      <b>
        LEI:
      </b>
       LEI123456
    </div>
    <div>
      <b>
        Indirizzo:
      </b>
       Via Roma, 1
    </div>
    <div>
      <b>
        Email:
      </b>
       info@azienda.it
    </div>
    <div>
      <b>
        Nazione:
      </b>
       Italia
    </div>
    <div>
      <b>
        Creatori:
      </b>
       1
    </div>
    <div>
      <b>
        Operatori:
      </b>
       1
    </div>
    <div>
      <b>
        Macchinari:
      </b>
       1
    </div>
    <div
      class="text-xs text-gray-400"
    >
      Creato il: 2025-01-01T12:00:00Z
    </div>
  </div>
</DocumentFragment>
`;
